os/mm/mm_heap/mm_initialize.c

```diff
@@ -146,6 +151,55 @@ void mm_addregion(FAR struct mm_heap_s *heap, FAR void *heapstart, size_t heapsi
 	heapinfo_update_node((FAR struct mm_allocnode_s *)heap->mm_heapstart[IDX], 0xDEAD);
 #endif
 
+#ifdef CONFIG_SMALL_MEMORY_ISOLATION
+	/* Divide one large chunk of memory into two chuncks of memory.
+	 * One is dedicated to being used for allocating small memory
+	 * and the other one is used for allocating larger memeory
+	 */
+
+	/* This memory block is the first one for small memory segments */
+	node			= (FAR struct mm_freenode_s *)(heapbase + SIZEOF_MM_ALLOCNODE);
+	node->size		= seg_size;
+	node->preceding = SIZEOF_MM_ALLOCNODE;
+	mm_addfreechunk(heap, node);
+	remain -= seg_size;
+
+	while (remain >= seg_size) {
+		node			= (FAR struct mm_freenode_s *)(((char *)node) + seg_size);
+		node->size		= seg_size;
+		node->preceding = seg_size;
+		remain -= seg_size;
+		mm_addfreechunk(heap, node);
+	}
+
+	/* There should be a delimiter between the isolated space for small memory segments
+	 * and the other one. This delimiter will be used to block merging two adjacent
+	 * memory segments across this delimiter.
+	 */
+	node			= (FAR struct mm_freenode_s *)(((char *)node) + seg_size);
+	node->size		= SIZEOF_MM_ALLOCNODE;
+	node->preceding = seg_size | MM_ALLOC_BIT;
+	heap->mm_delimiter = (FAR struct mm_allocnode_s *)node;
+#ifdef CONFIG_DEBUG_MM_HEAPINFO
+	/* Fill magic number 0xDEADDEAD as malloc info */
+	heapinfo_update_node((FAR struct mm_allocnode_s *)heap->mm_delimiter, 0xDEADDEAD);
+#endif
+
+	/* This memory block is the second one for larger memory segments */
```

Comment:
A : from 188 line to 200 line
B : from 203 line to 213 line,
A and B are almost same without the calculation of node->size.
How about unifying?

Reply:
At free-ing, MM tries to coalesce adjacent free nodes in the physical memory space, here 'prev' and 'next', with the released node. This is why almost the same logic performs twice. So it's hard to combine these two into one single logic. Let me consider more.